README.md
# [tnc-cup](https://cup.talkncloud.com)

[![GitHub license](https://img.shields.io/github/license/talkncloud/tnc-cup-client?color=blue&style=flat-square)](https://github.com/talkncloud/tnc-cup-client/blob/master/LICENSE)
[![npm version](https://img.shields.io/npm/v/@talkncloud/tnc-cup?style=flat-square)](https://www.npmjs.com/package/@talkncloud/tnc-cup)
[![Known Vulnerabilities](https://img.shields.io/snyk/vulnerabilities/github/talkncloud/tnc-cup-client?style=flat-square&targetFile=package.json)](https://snyk.io/test/github/talkncloud/tnc-cup-client?targetFile=package.json)


**Note**: This is currently in **PREVIEW**. There will be bugs, we will fix the bugs. 

talkncloud checkup (tnc-cup) has been developed by [talkncloud](https://www.talkncloud.com) and is a client for AWS price estimation. This cli client will parse cloudformation templates and post revelvant sections to the tnc-cup backend api to provide a price estimate.

Some key features:
  * Parse cloudformation & CDK (transpiled CF out)
  * Minimize post data by configuring client to remove key sections
  * Convert the estimate into your local currency
  * Support for multiple regions
  * Set a budget threshold
  * CI/CD friendly with error 0, 1, 2
  * Cli output with tabular display

## How to Install ##

```
npm install -g @talkncloud/tnc-cup
```

tnc-cup can now be used from the terminal simply by running "tnc-cup"

## Configuration ##

The local configuration contains common configuration parameters required to use tnc-cup. The configuration file is located in the $HOME directory named ".tnc-cup.config.json".

Install the sample config file by running the following command

```
tnc-cup -c
```

a new file will be create in $HOME/.tnc-cup.config.example.json

## Configuration Parameters ##

| Key | Type | Possible values | Description |
|---|---|---|---|
| api | object | | Required |
| api.baseUrl | string | https://api.talkncloud.com | Required, fixed |
| api.version | string | `client` | Required, fixed |
| api.header | object | | Required |
| api.header.x-api-key | string | `some api key` | Required | 
| templateEndpoint | string | `template` | Required, where data is posted to  - /template |
| servicesEndpoint | string | `services` | Required, list of supported services  - /services |
| find | object | |
| find.include | array | `["AWS::CDK::Metadata"]` | if include empty or undefined, it will show all resources |
| find.exclude | array | `["AWS::CDK::Metadata"]` | exclude will override include |
| other | array | [] | placeholder for future implementation |
| currency | string | `USD` | three letter country code to convery currency |
| budget | float | `50.00` | budget allocated for this template |
| region | string | `ap-southeast-2` | AWS compatible region code |


## API KEY ##

Go to [tnc-cup](https://cup.talkncloud.com/signup) to register, once registered use the dashboard to generate an API KEY. Use this key to update your tnc-cup.config.json.

## Cli parameters ##

1. -c = generate configuration file
2. -t = template file to parse

## Example ##

1. Navigate to your directory with the cloudformation template
2. tnc-cup -t mytemplate.json | yaml

<img src="https://cup.talkncloud.com/img/tnc-cup-client-sample.f731cf0c.png" alt="alt text" title="example output" />


## API ##
tnc-cup consists of a client that interfaces with the backend api. The backend api exposes a few endpoints and can be queried using typical curl tools. 

### Authorization ###
In general, to use the API you'll need an API key, the keys have typical rate limiting etc enabled. To use the endpoints below add the following to the header:

`x-api-key: 'YOUR AWESOME KEY HERE'`

### Endpoints ###

**Post body to api**
```http
POST /template
```
> Body
```javascript
[{"file":[{"WebInstance":{"Type":"AWS::EC2::Instance","Properties":{"BlockDeviceMappings":[{"DeviceName":"/dev/sda1","Ebs":{"VolumeSize":24000,"VolumeType":"gp3"}}],"InstanceType":"t2.nano","ImageId":"ami-80861296","KeyName":"my-key","Monitoring":true,"SecurityGroupIds":[{"class":"Ref","name":"Ref","data":"WebSecurityGroup"}],"SubnetId":"subnet-abc01234","Tags":[{"Key":"Name","Value":"webserver"}],"Volumes":[{"Device":"/dev/sdf","VolumeId":{"class":"Ref","name":"Ref","data":"LogVolume"}}]}}}],"currency":{"code":"USD"},"budget":10, "region": "us-east-1"}]
```
> Response
```javascript
[
    {
        "EC2": {
            "Compute": {
                "description": "$0.0058 per On Demand Linux t2.nano Instance Hour - ami: Canonical, Ubuntu, 16.04 LTS, amd64 xenial image build on 2017-04-14",
                "price": 4.0
            },
            "Storage-AMI": {
                "description": "$0.10 per GB-month of General Purpose SSD (gp2) provisioned storage - US East (Northern Virginia) size: 8GB",
                "price": 0.8
            },
            "Storage": {
                "description": "$0.005 per provisioned IOPS-month of gp3 - US East (N. Virginia) size: 24000GB",
                "price": 120.0
            }
        }
    },
    {
        "TOTAL (USD)": {
            "description": "Total estimated price (USD)",
            "price": 125
        }
    },
    {
        "budget": {
            "status": 1,
            "message": "over budget by $114 (USD)"
        }
    }
]
```
**List current supported services**
```http
GET /services
```
> Response
```javascript
{
    "AppSync:GraphQLApi": "AWS::AppSync::GraphQLApi",
    "Cognito:UserPool": "AWS::Cognito::UserPool",
    "EC2:Instance": "AWS::EC2::Instance",
    "EC2:Volume": "AWS::EC2::Volume",
    "RDS:DBInstance": "AWS::RDS::DBInstance",
    "S3:Bucket": "AWS::S3::Bucket",
    "WAFv2:WebACL": "AWS::WAFv2::WebACL"
}
```
**List supported currencies for conversion**
```http
GET /currency
```
> Response

```javascript
{
    "results": {
        "ALL": {
            "currencyName": "Albanian Lek",
            "currencySymbol": "Lek",
            "id": "ALL"
        },
        "XCD": {
            "currencyName": "East Caribbean Dollar",
            "currencySymbol": "$",
            "id": "XCD"
        },
        "EUR": {
            "currencyName": "Euro",
            "currencySymbol": "€",
            "id": "EUR"
        }...
```

### API response codes ###

| Status Code | Description |
| :--- | :--- |
| 200 | `OK` |
| 400 | `BAD REQUEST` |
| 403 | `ACCESS DENIED` |
| 429 | `TOO MANY REQUESTS / LIMIT` |
| 500 | `INTERNAL SERVER ERROR` |
| 502 | `BAD GATEWAY` |
| 503 | `UNAVAILABLE` |
| 504 | `TIMEOUT` |

### Contributing ###

Yes please! Check out our [contribution](https://github.com/talkncloud/tnc-cup-client/blob/main/CONTRIBUTING.md) guide.

CHANGELOG.md
# Changelog

## v0.0.19 (08/01/2021)

#### bug

- [**bug**] generate local config fails [#45](https://github.com/talkncloud/tnc-cup-client/issues/45)
- [**bug**] npm install error cli.js not found [#44](https://github.com/talkncloud/tnc-cup-client/issues/44)
- [**bug**] no parameter causes null object error [#43](https://github.com/talkncloud/tnc-cup-client/issues/43)
- [**bug**] services no longer valid [#33](https://github.com/talkncloud/tnc-cup-client/issues/33)
- [**bug**] unable to parse cloudformation yaml [#30](https://github.com/talkncloud/tnc-cup-client/issues/30)
- [**bug**] duplicate data sent from the client to the api [#13](https://github.com/talkncloud/tnc-cup-client/issues/13)
- [**bug**] "key" added to end of post payload [#7](https://github.com/talkncloud/tnc-cup-client/issues/7)
- [**bug**][**in progress**] cli only works when in the same directory as template [#1](https://github.com/talkncloud/tnc-cup-client/issues/1)

#### closed

- [**closed**] Npm build, publish [#38](https://github.com/talkncloud/tnc-cup-client/issues/38)
- [**closed**] new configuration item - region [#27](https://github.com/talkncloud/tnc-cup-client/issues/27)
- [**closed**] 403, 502 error handling [#23](https://github.com/talkncloud/tnc-cup-client/issues/23)

#### enhancement

- [**enhancement**] add remaining api status codes [#48](https://github.com/talkncloud/tnc-cup-client/issues/48)
- [**enhancement**] remove extra output [#47](https://github.com/talkncloud/tnc-cup-client/issues/47)
- [**enhancement**] Extend service col, shorten group col [#36](https://github.com/talkncloud/tnc-cup-client/issues/36)
- [**enhancement**] remove api-key from debug [#24](https://github.com/talkncloud/tnc-cup-client/issues/24)
- [**enhancement**] update output table colors [#17](https://github.com/talkncloud/tnc-cup-client/issues/17)
- [**enhancement**] format total price(s) to the right [#15](https://github.com/talkncloud/tnc-cup-client/issues/15)
- [**enhancement**] new configuration item - currency [#14](https://github.com/talkncloud/tnc-cup-client/issues/14)
- [**enhancement**] new configuration item - budget threshold [#11](https://github.com/talkncloud/tnc-cup-client/issues/11)
- [**enhancement**] format contents of result from api [#9](https://github.com/talkncloud/tnc-cup-client/issues/9)
- [**enhancement**][**wontfix**] Include / exclude sub keys [#4](https://github.com/talkncloud/tnc-cup-client/issues/4)
- [**enhancement**] Query api for supported services to filter template [#3](https://github.com/talkncloud/tnc-cup-client/issues/3)
- [**enhancement**][**in progress**] Return json result from api [#2](https://github.com/talkncloud/tnc-cup-client/issues/2)


index.ts
import fs from 'fs';
import Axios from 'axios';
import { Response } from './src/models/response';
import { Service } from './src/models/service-content';
import { BudgetStatus } from './src/models/budget-status';
import { BudgetResponse } from './src/models/budget-response';
import { parseYamlFromPath, constructTemplateBodyApi } from './src/utils/parser';
import Table from "cli-table3"
import colors from "colors";
import { Currency } from './src/models/currency';
import { terminateApp } from './src/utils/app';
import { TERMINATE_ON_ERROR } from './src/utils/constants';
const homedir = require('os').homedir();

function find(content: any, includeSearchContent: any[], excludeSearchContent?: any[], availableServices?: any[]) {
    // console.log(`find`);
    let arrayFindResult = [];
    for (let key of Object.keys(content['Resources'])) {
        const service = content['Resources'][key]['Type'];
        if (excludeSearchContent !== undefined
            && excludeSearchContent.length > 0
            && excludeSearchContent.includes(service)) {
            continue;
        }

        if (availableServices === undefined || availableServices === null) {
            continue;
        } else {
            if (availableServices.includes(service) &&
                (includeSearchContent === undefined
                    || includeSearchContent.length == 0
                    || (includeSearchContent !== undefined
                        && includeSearchContent.length > 0
                        && includeSearchContent.includes(service)))) {
                arrayFindResult.push({
                    [key]: content['Resources'][key]
                });
            }
        }
    }
    return arrayFindResult;
}

function proccessFile(
    cloudFormationFilePath: string,
    includeSearchContent: any[],
    excludeSearchContent?: any[],
    availableServices?: any[]
) {
    // console.log(`proccess ${cloudFormationFilePath}`);
    let response: Response = {
        data: [],
        message: ""
    };
    try {
        const cloudFormationContent = parseYamlFromPath(cloudFormationFilePath);
        const findResult = find(cloudFormationContent, includeSearchContent, excludeSearchContent, availableServices);
        // console.log(`Parse file: ${cloudFormationFilePath}`);
        if (findResult.length > 0) {
            response.data.push(...findResult);
        }
    } catch (e) {
        console.log(`Process with error ${e.message}`);
        response.message = "Not valid json or yaml file"
        throw response;
    }
    return response;
}

function proccessDirectory(
    cloudFormationDirPath: string,
    includeSearchContent: any[],
    excludeSearchContent?: any[],
    availableServices?: any[],
    additionalContentFromFilePath?: string
) {
    // console.log(`proccessDirectory ${cloudFormationDirPath}`);
    const path = require('path');
    const files = fs.readdirSync(cloudFormationDirPath);
    const arrayDirectoryResult = [];
    for (let file of files) {
        try {
            const findResult = proccessFile(path.join(cloudFormationDirPath, file), includeSearchContent, excludeSearchContent, availableServices);
            arrayDirectoryResult.push(...findResult.data);
        }
        catch (e) {
            console.log(`Process Dir with error ${e.message}`);
        }
    }
    return arrayDirectoryResult;
}

export async function proccessFromConfigFile(filePath: string, shouldShowJson: boolean, shouldOutputToFile: boolean) {
    // console.log(`proccessFromConfigFile ${filePath}`);
    try {
        const config = parseYamlFromPath(`${homedir}/.tnc-cup.config.json`);

        /*if((config.files === undefined || config.files.length == 0) 
            && (config.directories === undefined || config.directories.length == 0)) {
            console.log("files or directories required");
            return [];
        }*/

        if (config.url === undefined && config.url == "") {
            console.log("url required");
            return [];
        }
        const baseUrl = `${config.api.baseUrl}/${config.api.version}`;
        const availableServices = await Axios.get(`${baseUrl}/${config.api.servicesEndpoint}`, {
            headers: config.api.header
        });
        const services: any[] = Object.values(availableServices.data);
        // console.log(`Available Services: ${JSON.stringify(services)}`);

        let arrayResult = [];
        arrayResult.push(...proccessFile(filePath, config.find.include, config.find.exclude, services).data);
        /*for (let directory of config.directories) {
            arrayResult.push(...proccessDirectory(directory, config.find.include, config.find.exclude, services));
        }

        for (let file of config.files) {
            arrayResult.push(...proccessFile(file, config.find.include, config.find.exlude, services).data);
        }*/


        if (availableServices && availableServices.data) {
            const c: Currency = { code: `${config.currency}` };
            //console.log(`url: ${baseUrl}/${config.api.templateEndpoint}`);
            // console.log(`header: ${JSON.stringify(config.api.header)}`);
            //console.log(`body: ${constructTemplateBodyApi(arrayResult, c, config.budget, config.region)}`);
            // console.log(`data: ${JSON.stringify(arrayResult, null, 2)}`);

            const apiReturn = await Axios.post(
                `${baseUrl}/${config.api.templateEndpoint}`,
                constructTemplateBodyApi(arrayResult, c, config.budget, config.region),
                {
                    headers: config.api.header
                }
            );

            if (shouldShowJson) {
                console.log(`returned data: ${JSON.stringify(apiReturn.data, null, 2)}`);
            }

            const table = new Table({
                head: [colors.white("Service"), colors.white("Group"), colors.white("Description"), colors.white("Price")],
                colWidths: [15, 15, 90, 10],
                wordWrap: true,
                chars: {
                    'top': '═', 'top-mid': '╤', 'top-left': '╔', 'top-right': '╗'
                    , 'bottom': '═', 'bottom-mid': '╧', 'bottom-left': '╚', 'bottom-right': '╝'
                    , 'left': '║', 'left-mid': '╟', 'mid': '─', 'mid-mid': '┼'
                    , 'right': '║', 'right-mid': '╢', 'middle': '│'
                }
            })

            const a: any[] = [];
            let budgetResponse: BudgetResponse | null = null;
            apiReturn.data.forEach((e: any | Service) => {
                // console.log(`e --> ${JSON.stringify(e)}`);
                let group = '';
                let service = '';
                let description = '';
                let price = '';
                let isGroup = false;
                for (var item in e) {
                    // console.log(`Single Service: --> ${item}`);
                    for (var key in e[item]) {
                        // console.log(`key: --> ${key}`);
                        if (key.match(/description|price/)) {
                            service = item;
                            if (key === 'description') description = e[item][key];
                            if (key === 'price') price = e[item][key];
                            isGroup = false;
                        }
                        else {
                            isGroup = true;
                            continue;
                        }
                    }
                    if (!isGroup) {
                        // console.log(`push: --> ${group} ${service} ${description} ${price}`);
                        // Align description right if TOTAL
                        if (item.includes('TOTAL ')) {
                            table.push([colors.yellow(service), colors.yellow(group), { hAlign: 'right', content: colors.yellow(description) }, { hAlign: 'right', content: colors.yellow('$' + `${price}`) }]);
                        } else {
                            table.push([colors.green(service), colors.green(group), colors.green(description), { hAlign: 'right', content: colors.green('$' + `${price}`) }]);
                        }
                    } else {
                        group = item;
                        // console.log(`Group Service: --> ${item}`);

                        if (item.toLowerCase().includes('budget')) {
                            budgetResponse = e[item];
                        } else {
                            for (var subItem in e[item]) {
                                // console.log(`single service key: --> ${subItem}`);
                                service = subItem;
                                for (var key in e[item][subItem]) {
                                    // console.log(`subItem key key: --> ${key}`);
                                    if (key.match(/description|price/)) {
                                        if (key === 'description') description = e[item][subItem][key];
                                        if (key === 'price') price = e[item][subItem][key];
                                    }
                                }
                                // console.log(`push: --> ${group} ${service} ${description} ${price}`);
                                table.push([colors.green(service), colors.green(group), colors.green(description), { hAlign: 'right', content: colors.green('$' + `${price}`) }]);
                            }
                        }
                    }
                }

                a.push(e);
            });
            console.log('\n');
            console.log(table.toString());

            if (budgetResponse !== null) {
                switch ((budgetResponse as BudgetResponse).status) {
                    case BudgetStatus.NORMAL:
                        console.log(`\n\x1b[32mBudget: ${(budgetResponse as BudgetResponse).message}`);
                        break;
                    case BudgetStatus.WARNING:
                        console.log(`\n\x1b[33mBudget: ${(budgetResponse as BudgetResponse).message}`);
                        break;
                    case BudgetStatus.ERROR:
                        console.log(`\n\x1b[31mBudget: ${(budgetResponse as BudgetResponse).message}`);
                        break;
                    default:
                        console.log(`\n\x1b[33mBudget: Unknown message.`);
                        break;
                }
            }
            console.log('\n');
            return apiReturn;
        } else {
            return [];
        }

    } catch (e) {
        // if the url is incorrect it won't have a response object
        if (e.response) {
            if (e.response.status === 400) {
                console.log(`\n\x1b[31mBad request.`);
            }   
                else if (e.response.status === 403) {
                console.log(`\n\x1b[31mAccess denied: Please check your API key in config file.`);
            }
                else if (e.response.status === 429) {
                    console.log(`\n\x1b[31mToo many requests or limit reached.`);
            }
                else if (e.response.status === 500) {
                console.log(`\n\x1b[31mInternal error, please contact administrator.`);
            }
                else if (e.response.status === 502) {
                console.log(`\n\x1b[31mBad gateway, please contact administrator.`);
            }
                else if (e.response.status === 503) {
                console.log(`\n\x1b[31mUnavailable, please try again.`);
            }
                else if (e.response.status === 504) {
                console.log(`\n\x1b[31mTimeout, please try again.`);
            } else {
                console.log(`\n\x1b[31m${e.message}`);
            }
        
        } else {
            console.log(`\n\x1b[31m${e.message}`);
        }
        terminateApp(TERMINATE_ON_ERROR);
    }
}